File: Examples/Control/SMC/SMCExampleImplicit.py
#!/usr/bin/env python

# Siconos-sample, Copyright INRIA 2005-2012.
# Siconos is a program dedicated to modeling, simulation and control
# of non smooth dynamical systems.
# Siconos is a free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
# Siconos is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with Siconos; if not, write to the Free Software
# Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
#
# Contact: Vincent ACARY, siconos-team@lists.gforge.fr

from siconos.kernel import FirstOrderLinearDS, Model, TimeDiscretisation, \
    TimeStepping, ZeroOrderHoldOSI
from siconos.control.simulation import ControlManager
from siconos.control.sensor import LinearSensor
from siconos.control.controller import LinearSMC
from matplotlib.pyplot import subplot, title, plot, grid, show, xlabel, ylabel
from numpy import eye, empty, zeros, savetxt
from math import ceil
from matplotlib import rc
import matplotlib.pyplot as plt
import numpy as np
import scipy
from scipy import arange
rc('text', usetex=True)

# variable declaration
ndof = 2  # Number of degrees of freedom of your system
t0 = 0.0  # start time
T = 1  # end time
h = 1.0e-4  # time step for simulation
hControl = 1.0e-2  # time step for control
Xinit = 1.0  # initial position
theta = 0.5
N = ceil((T-t0)/h + 10)  # number of time steps
outputSize = 5  # number of variable to store at each time step

# Matrix declaration
A = zeros((ndof, ndof))
x0 = [Xinit, -Xinit]
sensorC = eye(ndof)
Csurface = [[0, 1]]
Brel = [[0], [2]]
#Drel = [[0, 0]]
# Simple check
if h > hControl:
    print "hControl must be bigger than h"
    exit(1)

# Declaration of the Dynamical System
processDS = FirstOrderLinearDS(x0, A)
# Model
process = Model(t0, T)
process.nonSmoothDynamicalSystem().insertDynamicalSystem(processDS)
# time discretisation
processTD = TimeDiscretisation(t0, h)
tSensor = TimeDiscretisation(t0, hControl)
tActuator = TimeDiscretisation(t0, hControl)
# Creation of the Simulation
processSimulation = TimeStepping(processTD, 0)
processSimulation.setName("plant simulation")
# Declaration of the integrator
processIntegrator = ZeroOrderHoldOSI(processDS)
processSimulation.insertIntegrator(processIntegrator)
# Actuator, Sensor & ControlManager
control = ControlManager(process)
sens = LinearSensor(tSensor, processDS, sensorC)
control.addSensorPtr(sens)
act = LinearSMC(tActuator, processDS)
act.setCsurfacePtr(Csurface)
act.setBPtr(Brel)
#act.setDPtr(Drel)
act.addSensorPtr(sens)
control.addActuatorPtr(act)

# Initialization
process.initialize(processSimulation)
control.initialize()
# This is not working right now
#eventsManager = s.eventsManager()

# Matrix for data storage
dataPlot = empty((3*(N+1), outputSize))
#dataPlot[0, 0] = processDS.t0()
dataPlot[0, 0] = t0
dataPlot[0, 1] = processDS.x()[0]
dataPlot[0, 2] = processDS.x()[1]
dataPlot[0, 3] = processDS.z()[0]
dataPlot[0, 4] = processDS.z()[1]


# Main loop
k = 1
while(processSimulation.hasNextEvent()):
    processSimulation.computeOneStep()
    dataPlot[k, 0] = processSimulation.nextTime()
    dataPlot[k, 1] = processDS.x()[0]
    dataPlot[k, 2] = processDS.x()[1]
    dataPlot[k, 3] = processDS.z()[0]
    dataPlot[k, 4] = processDS.z()[1]
    k += 1
    processSimulation.nextStep()
# Resize
dataPlot.resize(k, outputSize)
# Save to disk
savetxt('SMCExampleImplicit-py.dat', dataPlot)
# Plot interesting data

#subplot(411)
#title('x1')
#plot(dataPlot[:, 0], dataPlot[:, 1])
#grid()
#subplot(412)
#title('x2')
#plot(dataPlot[:, 0], dataPlot[:, 2])
#grid()
#subplot(413)
#title('u1')
#plot(dataPlot[:, 0], dataPlot[:, 3])
#grid()
#subplot(414)
#title('u2')
#plot(dataPlot[:, 0], dataPlot[:, 4])
#show()

subplot(211)
ylabel(r'$\sigma$')
xlabel(r't')
plot(dataPlot[:, 0], dataPlot[:, 2])
grid()
subplot(212)
ylabel(r'$\bar{u}^s$')
xlabel(r't')
plt.ylim(-2.1, 2.1)
plot(dataPlot[:, 0], dataPlot[:, 4])
show()

subplot(211)
ylabel(r'$\sigma$')
xlabel(r't')
plt.xlim(xmin=.49)
plt.ylim(-0.03, 0.03)
plot(dataPlot[4900:, 0], dataPlot[4900:, 2])
grid()
subplot(212)
ylabel(r'$\bar{u}^s$')
xlabel(r't')
plt.ylim(-2.1, 2.1)
plt.xlim(xmin=.49)
p1 = plot(dataPlot[4900:, 0], dataPlot[4900:, 4])
#p2 = plot(dataPlot[4900:, 0], np.sin(50*dataPlot[4900:, 0]))
#plt.legend((p1[0], p2[0]), (r'$\bar{u}^s(t)$', r'$-\rho(t)$'), ncol=2)
show()

u_z = dataPlot[5100:, 4]
n = len(u_z)
Y = scipy.fft(dataPlot[5100:, 4])/n
k = arange(n)
T = n*h
frq = k/T
frq = frq[range(n/2)]
Y = Y[range(n/2)]
plot(frq, abs(Y), 'r')
xlabel(r'freq (Hz)')
title(r'Frequency spectrum of $\bar{u}^s$')
show()

# TODO
# compare with the reference
#ref = getMatrix(SimpleMatrix("result.ref"))
#if (norm(dataPlot - ref[1:,:]) > 1e-12):
#    print("Warning. The result is rather different from the reference file.")

File: examples/Control/SMC/SMCExampleImplicit.cpp
/* Siconos is a program dedicated to modeling, simulation and control
 * of non smooth dynamical systems.
 *
 * Copyright 2018 INRIA.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
*/

/* !\file SMCExampleImplicit.cpp
  \brief Two independent systems of dimension one controlled to slide
  on \f$x = 0\f$. An implicit scheme is used
  O. Huber
  */

#include "SiconosKernel.hpp"
#include "SiconosControl.hpp"
using namespace std;
class MyDS : public FirstOrderLinearDS
{
public:
  MyDS(SP::SiconosVector x0, SP::SiconosMatrix A) : FirstOrderLinearDS(x0, A)
  {
    _b.reset(new SiconosVector(x0->size()));
  };
  void computeb(double time)
  {
    printf("computeB\n");
    double t = sin(50 * time);
    _b->setValue(0,t);
    _b->setValue(1,-t);
    printf("b[0] = %g, b[1] = %g\n", _b->getValue(0), _b->getValue(1));
  };
};
// main program
int main(int argc, char* argv[])
{
  // User-defined parameters
  unsigned int ndof = 2;          // Number of degrees of freedom of your system
  double t0 = 0.0;                // Starting time
  double T = 1;                   // Total simulation time
  double h = 1.0e-4;              // Time step for simulation
  double hControl = 1.0e-2;       // Time step for control
  double Xinit = 1.0;

  if (h > hControl)
  {
    RuntimeException::selfThrow("hControl must be bigger than h");
  }

  // ================= Creation of the model =======================
  // Steps:
  // - create a Dynamical System
  // - add a Simulation to the model

  // -------------------------
  // --- Dynamical systems ---
  // -------------------------

  // First System:
  // dx/dt = Ax + u(t) + r
  // x(0) = x0
  // Note: r = Blambda, B defines in relation below.

  // Matrix declaration
  SP::SiconosMatrix A(new SimpleMatrix(ndof, ndof, 0));
  SP::SiconosVector x0(new SiconosVector(ndof));
  (*x0)(0) = Xinit;
  (*x0)(1) = -Xinit;
  SP::SimpleMatrix sensorC(new SimpleMatrix(2, 2));
  sensorC->eye();
  SP::SimpleMatrix sensorD(new SimpleMatrix(2, 2, 0));
  SP::SimpleMatrix Csurface(new SimpleMatrix(1, 2, 0));
  (*Csurface)(0, 1) = 1;
  SP::SimpleMatrix Brel(new SimpleMatrix(2, 1, 0));
  (*Brel)(1, 0) = 2;

  // Dynamical Systems
  SP::FirstOrderLinearDS processDS(new MyDS(x0, A));
  // -------------
  // --- Model process ---
  // -------------
  SP::ControlSimulation sim(new ControlZOHSimulation(t0, T, h));
  //sim->setSaveOnlyMainSimulation(true);
  sim->addDynamicalSystem(processDS);

  // ------------------
  // --- Simulation ---
  // ------------------
  // Control stuff
  // use a controlSensor
  SP::LinearSensor sens(new LinearSensor(processDS, sensorC, sensorD));
  sim->addSensor(sens, hControl);
  // add the sliding mode controller
  SP::LinearSMC act(new LinearSMC(sens));
  act->setCsurface(Csurface);
  act->setB(Brel);
  sim->addActuator(act, hControl);
  // =========================== End of model definition ===========================

  // ================================= Computation =================================

  // --- Simulation initialization ---

  cout << "====> Simulation initialisation ..." << endl << endl;
  // initialise the process and the ControlManager
  sim->initialize();

  // ==== Simulation loop =====
  cout << "====> Start computation ... " << endl << endl;
  sim->run();
  // --- Output files ---
  cout << "====> Output file writing ..." << endl;
  SimpleMatrix& dataPlot = *sim->data();
  ioMatrix::write("SMCExampleImplicit.dat", "ascii", dataPlot, "noDim");

  // // Comparison with a reference file
  // SimpleMatrix dataPlotRef(dataPlot);
  // dataPlotRef.zero();
  // ioMatrix::read("SMCExampleImplicit.ref", "ascii", dataPlotRef);
  // std::cout << (dataPlot - dataPlotRef).normInf() << std::endl;

  // if ((dataPlot - dataPlotRef).normInf() > 1e-12)
  // {
  //   std::cout << "Warning. The results is rather different from the reference file." << std::endl;
  //   return 1;
  // }

}
